### ISSUE TITLE
Port FxCop rule CA2000: DisposeObjectsBeforeLosingScope

### ISSUE THREAD

**Issue by lgolding** [dotnet/roslyn-analyzers#476]
**Title:** Dispose Objects Before Losing Scope

**Description:**

Because an exceptional event might occur that will prevent the finalizer of an object from running, the object should be explicitly disposed before all references to it are out of scope.

**Dependency:** Dataflow

**Notes:**

**Comment by srivatsn** [comment]
Blocked on dataflow analysis

**Comment by alaforte** [comment]
I would be grateful if you will port this rule. Thank you

**Comment by BenniBuer** [comment]
Would be great to have this rule. Is there an update on the status of this?

**Comment by gafter** [comment]
CA2000 was implemented in the C# compiler as a proof of concept some time ago. We removed it from the compiler as the intent is to expand the scope of what analyzers can do to cover these kinds of use cases:

- https://github.com/dotnet/roslyn/commit/8474c254666eee18c633601a789f8d41b685b51c
- https://github.com/dotnet/roslyn/commit/76911d6f85c03f3d068a5ea45530638ab2ce368a

I believe our intent is that the IOperation APIs should eventually be capable of expressing this kind of rule. But we still have a long way to go before that is possible.

**Comment by mavasani** [comment]
Need to ensure that issues mentioned in https://github.com/dotnet/roslyn-analyzers/issues/1404 are addressed when this is implemented.

**Comment by alaforte** [comment]
Sorry, I didn't got the point

Thank you

2017-11-27 18:00 GMT+01:00 Manish Vasani <notifications@github.com>:

**Comment by drauch** [comment]
What's the status on this one? : ) Since we've ported our projects to the new format we cannot use FxCop anymore - we still want to get warned when not disposing objects! Is there a workaround so that the compiler outputs a warning? Some kind of analyzer? At least a beta analyzer?

**Comment by sharwell** [comment]
:memo: In the meantime, the [DotNetAnalyzers/IDisposableAnalyzers](https://github.com/DotNetAnalyzers/IDisposableAnalyzers) project by @JohanLarsson may provide the desired functionality.

**Comment by mavasani** [comment]
@drauch We are currently working on preliminary DFA based experimental analyzer for dispose rules. We hope to have a beta package with the experimental DFA rules published in a few weeks.

**Comment by drauch** [comment]
Thanks @sharwell , this looks really nice at first sight. I'll give it a try!

**Comment by mavasani** [comment]
A PR for preliminary dataflow analysis based implementation of this rule is out for review: https://github.com/dotnet/roslyn-analyzers/pull/1583

As mentioned in the PR, following are the TODOs to bring it to parity with original FxCop implementation:
   1. Analyze exceptional code paths (#1580)
   2. Path sensitive analysis (#1581)
   3. Improve preciseness by moving to compiler CFG (#1567)

We hope to address these TODOs soon.

**Comment by mavasani** [comment]
This rule has been implemented in the experimental DFA package. https://github.com/dotnet/roslyn-analyzers/pull/1583#issue-169473457 lists the TODO items, each of which are tracked with separate issues. Marking this issue as complete.
